File: src/steps.js
'use strict';

// Step / cadence detection from the Polar H10's chest accelerometer.
//
// A chest strap picks up the vertical bounce of each footfall, so a streaming
// pedometer works on the same 25 Hz ACC samples we use for posture:
//   1. acceleration magnitude |a| (mg)
//   2. remove gravity with a slow EMA baseline  -> dynamic component
//   3. light low-pass, then detect local maxima above an adaptive threshold
//      within the walking cadence window (≈ 0.25–1.4 s between steps)
//   4. require a couple of rhythmic peaks in a row before counting, so single
//      jolts (typing, gestures) don't register as steps.
// Chest mounting is less accurate than wrist/hip — good for cadence and
// "walking vs not", rough (±10 %ish) for absolute counts.

function median(arr) {
  if (!arr.length) return 0;
  const a = arr.slice().sort((x, y) => x - y);
  const m = a.length >> 1;
  return a.length % 2 ? a[m] : (a[m - 1] + a[m]) / 2;
}

class StepCounter {
  constructor({ sampleRate = 25 } = {}) {
    this.dt = 1 / sampleRate;
    this.aGrav = 1 - Math.exp(-1 / (1.0 * sampleRate)); // gravity baseline (~1 s)
    this.aLp = 1 - Math.exp(-1 / (0.08 * sampleRate));  // light smoothing
    this.grav = null;
    this.lp = 0; this.prevLp = 0; this.rising = false;
    this.tSec = 0; this.lastPeakSec = -10;
    this.env = 0;            // adaptive peak-amplitude envelope (mg)
    this.run = 0;            // consecutive in-cadence peaks
    this.intervals = [];     // recent step intervals (s) for cadence
    this.steps = 0;          // cumulative steps since construction
    this.lastStepSec = -10;  // signal-clock time of last counted step
  }

  static get MIN_INT() { return 0.25; }  // ≤ 240 spm (reject double-count)
  static get MAX_INT() { return 1.4; }   // ≥ ~43 spm
  static get MIN_AMP() { return 80; }    // mg, floor for the adaptive threshold
  static get IDLE_SEC() { return 2.5; }  // no step within this → not walking

  add(s) {
    if (!s) return;
    const m = Math.sqrt(s.x * s.x + s.y * s.y + s.z * s.z);
    if (this.grav == null) { this.grav = m; this.lp = 0; this.prevLp = 0; this.tSec = 0; return; }
    this.grav += this.aGrav * (m - this.grav);
    const dyn = m - this.grav;
    this.prevLp = this.lp;
    this.lp += this.aLp * (dyn - this.lp);
    this.tSec += this.dt;

    const wasRising = this.rising;
    this.rising = this.lp > this.prevLp;
    if (wasRising && !this.rising) {            // local maximum just passed
      const amp = this.prevLp;
      const thr = Math.max(StepCounter.MIN_AMP, 0.5 * this.env);
      if (amp > thr) {
        this.env += 0.3 * (amp - this.env);
        const interval = this.tSec - this.lastPeakSec;
        if (interval >= StepCounter.MIN_INT && interval <= StepCounter.MAX_INT) {
          this.run++;
          this.intervals.push(interval);
          if (this.intervals.length > 6) this.intervals.shift();
          if (this.run === 2) this.steps += 2;       // confirm rhythm: count the first pair
          else if (this.run > 2) this.steps += 1;
          this.lastStepSec = this.tSec;
        } else {
          this.run = 0; this.intervals.length = 0;    // rhythm broken
        }
        this.lastPeakSec = this.tSec;
      }
    }
  }

  walking() { return (this.tSec - this.lastStepSec) < StepCounter.IDLE_SEC; }

  // Steps per minute from recent intervals, or 0 if not currently walking.
  cadence() {
    if (!this.walking() || this.intervals.length < 2) return 0;
    const med = median(this.intervals);
    return med > 0 ? Math.round(60 / med) : 0;
  }

  snapshot() {
    return { steps: this.steps, cadence: this.cadence(), walking: this.walking() };
  }
}

module.exports = { StepCounter, median };

// ---- self-test: `node src/steps.js` --------------------------------------
if (require.main === module) {
  let fail = 0;
  const ok = (cond, msg) => { if (!cond) { console.error(`FAIL ${msg}`); fail++; } };
  const SR = 25;

  // 10 s of walking at 2 Hz (≈120 spm) → ~20 steps. Vertical bounce ±200 mg on
  // top of 1000 mg gravity; small noise on the other axes.
  const sc = new StepCounter({ sampleRate: SR });
  const N = 10 * SR;
  for (let i = 0; i < N; i++) {
    const t = i / SR;
    const z = -1000 - 200 * Math.sin(2 * Math.PI * 2 * t);
    sc.add({ x: 12 * Math.sin(2 * Math.PI * 2 * t + 1), y: 8 * Math.random(), z });
  }
  ok(sc.steps >= 16 && sc.steps <= 24, `~20 steps walking (got ${sc.steps})`);
  ok(sc.cadence() >= 110 && sc.cadence() <= 130, `~120 spm cadence (got ${sc.cadence()})`);

  // 6 s sitting still (tiny noise) → no new steps.
  const before = sc.steps;
  for (let i = 0; i < 6 * SR; i++) sc.add({ x: 3 * (Math.random() - .5), y: 3 * (Math.random() - .5), z: -1000 + 3 * (Math.random() - .5) });
  ok(sc.steps === before, `no steps while still (added ${sc.steps - before})`);
  ok(sc.cadence() === 0, `cadence 0 when idle (got ${sc.cadence()})`);

  console.log(fail === 0 ? 'steps.js self-test: OK' : `steps.js self-test: ${fail} FAILED`);
  process.exit(fail === 0 ? 0 : 1);
}

File: app/android/app/src/main/java/dev/otake/rmssdh10n/hrv/Steps.java
package dev.otake.rmssdh10n.hrv;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Step / cadence detection from the H10's chest accelerometer. Java port of
 * src/steps.js: gravity-removed dynamic magnitude, light low-pass, adaptive
 * peak detection within the walking cadence window, and a rhythm gate so single
 * jolts don't count. Liveness uses the signal clock (tSec from ACC samples),
 * not wall time, so it behaves identically off-screen.
 */
public final class Steps {
    static final double MIN_INT = 0.25; // <= 240 spm (reject double-count)
    static final double MAX_INT = 1.4;  // >= ~43 spm
    static final double MIN_AMP = 80;   // mg, floor for adaptive threshold
    static final double IDLE_SEC = 2.5; // no step within this -> not walking

    private final double dt, aGrav, aLp;
    private boolean hasGrav = false;
    private double grav = 0;
    private double lp = 0, prevLp = 0;
    private boolean rising = false;
    private double tSec = 0, lastPeakSec = -10;
    private double env = 0;
    private int run = 0;
    private final List<Double> intervals = new ArrayList<>();
    public int steps = 0;
    private double lastStepSec = -10;

    public Steps() { this(25); }

    public Steps(int sampleRate) {
        this.dt = 1.0 / sampleRate;
        this.aGrav = 1 - Math.exp(-1.0 / (1.0 * sampleRate));
        this.aLp = 1 - Math.exp(-1.0 / (0.08 * sampleRate));
    }

    public void add(double sx, double sy, double sz) {
        double m = Math.sqrt(sx * sx + sy * sy + sz * sz);
        if (!hasGrav) { grav = m; lp = 0; prevLp = 0; tSec = 0; hasGrav = true; return; }
        grav += aGrav * (m - grav);
        double dyn = m - grav;
        prevLp = lp;
        lp += aLp * (dyn - lp);
        tSec += dt;

        boolean wasRising = rising;
        rising = lp > prevLp;
        if (wasRising && !rising) { // local maximum just passed
            double amp = prevLp;
            double thr = Math.max(MIN_AMP, 0.5 * env);
            if (amp > thr) {
                env += 0.3 * (amp - env);
                double interval = tSec - lastPeakSec;
                if (interval >= MIN_INT && interval <= MAX_INT) {
                    run++;
                    intervals.add(interval);
                    if (intervals.size() > 6) intervals.remove(0);
                    if (run == 2) steps += 2;       // confirm rhythm: count the first pair
                    else if (run > 2) steps += 1;
                    lastStepSec = tSec;
                } else {
                    run = 0; intervals.clear();      // rhythm broken
                }
                lastPeakSec = tSec;
            }
        }
    }

    public boolean walking() { return (tSec - lastStepSec) < IDLE_SEC; }

    public int cadence() {
        if (!walking() || intervals.size() < 2) return 0;
        List<Double> s = new ArrayList<>(intervals);
        Collections.sort(s);
        int mid = s.size() >> 1;
        double med = (s.size() % 2 == 1) ? s.get(mid) : (s.get(mid - 1) + s.get(mid)) / 2.0;
        return med > 0 ? (int) Math.round(60 / med) : 0;
    }
}
